src/irmin-pack/unix/file_manager.ml

```diff
@@ -383,6 +383,30 @@ struct
     let path = Irmin_pack.Layout.V3.control ~root in
     Control.create_rw ~path ~overwrite pl
 
+  let upgrade_v3_to_v4 : Control_file.Payload_v3.t -> Payload.t =
+   fun pl ->
+    let status =
+      match pl.status with
+      | From_v1_v2_post_upgrade x -> Payload.From_v1_v2_post_upgrade x
+      | From_v3_no_gc_yet -> No_gc_yet
+      | From_v3_used_non_minimal_indexing_strategy ->
+          Used_non_minimal_indexing_strategy
+      | From_v3_gced x ->
+          Gced
+            {
+              suffix_start_offset = x.suffix_start_offset;
+              generation = x.generation;
+              oldest_live_commit_offset = x.suffix_start_offset;
+            }
+      | _ -> assert false
+    in
+    {
+      dict_end_poff = pl.dict_end_poff;
+      suffix_end_poff = pl.suffix_end_poff;
+      status;
+      upgraded_from_v3_to_v4 = true;
+    }
```

Comment:
Why can't this logic live in the control file since it doesn't depend on anything beyond its own scope?

Reply:
Ah yes true, `oldest_live_commit_offset = x.suffix_start_offset;` is an operation that is local to the control file. I told @icristescu to move the upgrade out of the CF because I expected a more complicated process that needed to do stuff with other files.

Let's move that back to CF

Reply:
Glad I didn't miss something here. Happy it can live in the CF!

Reply:
moved it back in CF